src/quak/_widget.py

```diff
@@ -37,7 +39,10 @@ def __init__(self, data, *, table: str = "df"):
             conn = data
         else:
             conn = duckdb.connect(":memory:")
-            if is_arrow_ipc(data):
+            if has_pycapsule_stream_interface(data):
+                # arrow_table = pa.RecordBatchReader.from_stream(data)
+                arrow_table = pa.table(data)
```

Comment:
Here this is not ideal because `pa.table` materializes the entire stream in memory at once. I tried with the `RecordBatchReader` but I think you might need to adjust your queries to actually _create a table_ instead of querying the stream directly. I assume you do multiple queries on the input... and presumably the _first query_ exhausts the input stream and then the second query has no data.

<img width="1361" alt="image" src="https://github.com/user-attachments/assets/8efbd2ab-7fd2-4685-99f7-9fe495432bb6">

Note that it shows 11k rows in the bottom right, but then doesn't display anything. So I think you need to persist the Arrow stream input manually.

Reply:
Yeah I figured this is not idea... I just wasn't sure how to wire up duckdb to query something consistently. I'm guessing something like "CREATE VIEW" from the input stream might allow us register some tables that duckdb can continuously query.

Reply:
I'm not 100% certain but I don't think you could use `CREATE VIEW` here for Arrow streams. Reading https://duckdb.org/docs/api/python/data_ingestion#directly-accessing-dataframes-and-arrow-objects that makes me think the first time you query the view it will exhaust the input stream.

I think we're running into the same issue as in ibis: https://github.com/ibis-project/ibis/pull/9663#issuecomment-2246268513 . In particular, the presence of the dunder method doesn't tell you whether the input data is already materialized or a stream. If you _know_ that the input data is an in-memory table, then you can call `__arrow_c_stream__` multiple times and get the full data each time. In that case you'd prefer to make a DuckDB view, because then you don't need to copy the input data. But if the input is a stream, then you would _need_ to persist it into a full duckdb table for ongoing queries.